flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBNativeMetricMonitor.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.contrib.streaming.state;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.metrics.Gauge;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.metrics.View;
+
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.GuardedBy;
+
+import java.io.Closeable;
+import java.math.BigInteger;
+
+/**
+ * A monitor which pulls {{@link RocksDB}} native metrics
+ * and forwards them to Flink's metric group. All metrics are
+ * unsigned longs and are reported at the column family level.
+ */
+@Internal
+public class RocksDBNativeMetricMonitor implements Closeable {
+	private static final Logger LOG = LoggerFactory.getLogger(RocksDBNativeMetricMonitor.class);
+
+	private final RocksDBNativeMetricOptions options;
+
+	private final MetricGroup metricGroup;
+
+	private final Object lock;
+
+	@GuardedBy("lock")
+	private RocksDB rocksDB;
+
+	RocksDBNativeMetricMonitor(
+		@Nonnull RocksDB rocksDB,
+		@Nonnull RocksDBNativeMetricOptions options,
+		@Nonnull MetricGroup metricGroup
+	) {
+		this.options = options;
+		this.metricGroup = metricGroup;
+		this.rocksDB = rocksDB;
+
+		this.lock = new Object();
+	}
+
+	/**
+	 * Register gauges to pull native metrics for the column family.
+	 * @param columnFamilyName group name for the new gauges
+	 * @param handle native handle to the column family
+	 */
+	void registerColumnFamily(String columnFamilyName, ColumnFamilyHandle handle) {
+		MetricGroup group = metricGroup.addGroup(columnFamilyName);
+
+		for (String property : options.getProperties()) {
+			RocksDBNativeMetricView gauge = new RocksDBNativeMetricView(handle, property);
+			group.gauge(property, gauge);
+		}
+	}
+
+	/**
+	 * Updates the value of metricView if the reference is still valid.
+	 */
+	private void setProperty(ColumnFamilyHandle handle, String property, RocksDBNativeMetricView metricView) {
+		if (metricView.isClosed()) {
```

Comment:
@StefanRRichter Do you know anything about the reliability of  `RocksDB#getLongProperty`?

Reply:
@zentol I did a little digging in the rocksdb code. The only place an exception that is explicitly thrown is in the JNI wrapper if the property string is invalid[1]. 

Otherwise, the value is just pulled from a map here[2]. Since there is already a test validating each property against the current rocksdb version, and this pr only exposes an enum, I don't think there is any reason to believe a RocksDB exception will be thrown in production. 

[1] https://github.com/facebook/rocksdb/blob/02dc07491650b0c8d8470a24806ba86d09ea31a4/java/rocksjni/rocksjni.cc#L1757

[2] https://github.com/facebook/rocksdb/blob/189f0c27aaecdf17ae7fc1f826a423a28b77984f/db/internal_stats.cc#L339

Reply:
cool, thanks for looking into it. We can leave the code as is.